File: apps/dsa-web/src/api/screener.ts
import apiClient from './index';
import { toCamelCase } from './utils';

export interface ScreenerFilterRequest {
  changePctMin?: number;
  changePctMax?: number;
  volumeRatioMin?: number;
  volumeRatioMax?: number;
  turnoverRateMin?: number;
  turnoverRateMax?: number;
  peRatioMin?: number;
  peRatioMax?: number;
  pbRatioMin?: number;
  pbRatioMax?: number;
  circMvMin?: number;
  circMvMax?: number;
  amplitudeMin?: number;
  amplitudeMax?: number;
  priceMin?: number;
  priceMax?: number;
  near52wHighPct?: number;
  near52wLowPct?: number;
  change60dMin?: number;
  change60dMax?: number;
}

export interface ScreenerRequest {
  preset?: string;
  filter?: ScreenerFilterRequest;
  sortBy?: string;
  sortDesc?: boolean;
  limit?: number;
}

export interface ScreenerResultItem {
  code: string;
  name?: string;
  price?: number;
  changePct?: number;
  changeAmount?: number;
  volume?: number;
  amount?: number;
  volumeRatio?: number;
  turnoverRate?: number;
  amplitude?: number;
  openPrice?: number;
  high?: number;
  low?: number;
  preClose?: number;
  peRatio?: number;
  pbRatio?: number;
  totalMv?: number;
  circMv?: number;
  change60d?: number;
  high52w?: number;
  low52w?: number;
}

export interface StrategyInfo {
  key: string;
  name: string;
  description: string;
}

export interface ScreenerResponse {
  results: ScreenerResultItem[];
  totalMatched: number;
  totalScanned: number;
  strategy?: StrategyInfo;
  scannedAt: string;
  cacheEmpty: boolean;
}

export interface PresetItem {
  key: string;
  name: string;
  description: string;
}

export interface PresetListResponse {
  presets: PresetItem[];
}

export const screenerApi = {
  getPresets: async (): Promise<PresetListResponse> => {
    const res = await apiClient.get('/api/v1/screener/presets');
    return toCamelCase<PresetListResponse>(res.data);
  },

  scan: async (req: ScreenerRequest): Promise<ScreenerResponse> => {
    const payload: Record<string, unknown> = {
      sort_desc: req.sortDesc ?? true,
      limit: req.limit ?? 50,
    };
    if (req.preset) payload.preset = req.preset;
    if (req.sortBy) payload.sort_by = req.sortBy;
    if (req.filter) {
      const f = req.filter;
      payload.filter = {
        change_pct_min: f.changePctMin,
        change_pct_max: f.changePctMax,
        volume_ratio_min: f.volumeRatioMin,
        volume_ratio_max: f.volumeRatioMax,
        turnover_rate_min: f.turnoverRateMin,
        turnover_rate_max: f.turnoverRateMax,
        pe_ratio_min: f.peRatioMin,
        pe_ratio_max: f.peRatioMax,
        pb_ratio_min: f.pbRatioMin,
        pb_ratio_max: f.pbRatioMax,
        circ_mv_min: f.circMvMin,
        circ_mv_max: f.circMvMax,
        amplitude_min: f.amplitudeMin,
        amplitude_max: f.amplitudeMax,
        price_min: f.priceMin,
        price_max: f.priceMax,
        near_52w_high_pct: f.near52wHighPct,
        near_52w_low_pct: f.near52wLowPct,
        change_60d_min: f.change60dMin,
        change_60d_max: f.change60dMax,
      };
    }
    const res = await apiClient.post('/api/v1/screener/scan', payload);
    return toCamelCase<ScreenerResponse>(res.data);
  },

  scanByPreset: async (presetKey: string, limit = 50): Promise<ScreenerResponse> => {
    const res = await apiClient.get(`/api/v1/screener/scan/${presetKey}`, {
      params: { limit },
    });
    return toCamelCase<ScreenerResponse>(res.data);
  },
};

File: api/v1/schemas/screener.py
# -*- coding: utf-8 -*-
"""
===================================
策略筛选 API Schema
===================================
"""

from typing import Any, Dict, List, Optional

from pydantic import BaseModel, Field


class ScreenerFilterRequest(BaseModel):
    """自定义筛选条件"""
    change_pct_min: Optional[float] = Field(None, description="涨跌幅下限(%)")
    change_pct_max: Optional[float] = Field(None, description="涨跌幅上限(%)")
    volume_ratio_min: Optional[float] = Field(None, description="量比下限")
    volume_ratio_max: Optional[float] = Field(None, description="量比上限")
    turnover_rate_min: Optional[float] = Field(None, description="换手率下限(%)")
    turnover_rate_max: Optional[float] = Field(None, description="换手率上限(%)")
    pe_ratio_min: Optional[float] = Field(None, description="PE下限")
    pe_ratio_max: Optional[float] = Field(None, description="PE上限")
    pb_ratio_min: Optional[float] = Field(None, description="PB下限")
    pb_ratio_max: Optional[float] = Field(None, description="PB上限")
    circ_mv_min: Optional[float] = Field(None, description="流通市值下限(元)")
    circ_mv_max: Optional[float] = Field(None, description="流通市值上限(元)")
    amplitude_min: Optional[float] = Field(None, description="振幅下限(%)")
    amplitude_max: Optional[float] = Field(None, description="振幅上限(%)")
    price_min: Optional[float] = Field(None, description="股价下限")
    price_max: Optional[float] = Field(None, description="股价上限")
    near_52w_high_pct: Optional[float] = Field(None, description="距52周新高百分比以内")
    near_52w_low_pct: Optional[float] = Field(None, description="距52周新低百分比以内")
    change_60d_min: Optional[float] = Field(None, description="60日涨跌幅下限(%)")
    change_60d_max: Optional[float] = Field(None, description="60日涨跌幅上限(%)")


class ScreenerRequest(BaseModel):
    """筛选请求"""
    preset: Optional[str] = Field(None, description="预设策略名称")
    filter: Optional[ScreenerFilterRequest] = Field(None, description="自定义筛选条件")
    sort_by: Optional[str] = Field(None, description="排序字段")
    sort_desc: bool = Field(True, description="是否降序")
    limit: int = Field(50, ge=1, le=200, description="返回条数上限")


class StrategyInfo(BaseModel):
    """策略信息"""
    key: str
    name: str
    description: str


class ScreenerResponse(BaseModel):
    """筛选响应"""
    results: List[Dict[str, Any]]
    total_matched: int
    total_scanned: int
    strategy: Optional[StrategyInfo] = None
    scanned_at: str
    cache_empty: bool = False


class PresetItem(BaseModel):
    """预设策略条目"""
    key: str
    name: str
    description: str


class PresetListResponse(BaseModel):
    """预设策略列表响应"""
    presets: List[PresetItem]
